mne/time_frequency/psd.py

```diff
@@ -112,6 +115,207 @@ def _check_nfft(n, n_fft, n_overlap):
+
+    time_mask = _time_mask(inst.times, tmin, tmax)
+    if picks is None:
+        picks = pick_types(inst.info, meg=True, eeg=True, ref_meg=False,
+                           exclude='bads')
+    if proj:
+        # Copy first so it's not modified
+        inst = inst.copy().apply_proj()
+
+    sfreq = inst.info['sfreq']
+    if isinstance(inst, _BaseRaw):
+        start, stop = np.where(time_mask)[0][[0, -1]]
+        data, times = inst[picks, start:(stop + 1)]
+    elif isinstance(inst, _BaseEpochs):
+        data = inst.get_data()[:, picks][..., time_mask]
+    elif isinstance(inst, Evoked):
+        data = inst.data[picks][..., time_mask]
+
+    return data, sfreq
+
+
+def _psd_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
+               n_jobs=1):
+    """Helper function for calculating Welch PSD."""
+    from scipy.signal import welch
+    dshape = x.shape[:-1]
+    n_times = x.shape[-1]
+    x = x.reshape(-1, n_times)
+
+    # Prep the PSD
+    n_fft, n_overlap = _check_nfft(n_times, n_fft, n_overlap)
+    win_size = n_fft / float(sfreq)
+    logger.info("Effective window size : %0.3f (s)" % win_size)
+    freqs = np.arange(n_fft // 2 + 1, dtype=float) * (sfreq / n_fft)
+    freq_mask = (freqs >= fmin) & (freqs <= fmax)
+    freqs = freqs[freq_mask]
+
+    # Parallelize across first N-1 dimensions
+    parallel, my_pwelch, n_jobs = parallel_func(_pwelch, n_jobs=n_jobs)
+    x_splits = np.array_split(x, n_jobs)
+    f_psd = parallel(my_pwelch(d, noverlap=n_overlap, nfft=n_fft,
+                     fs=sfreq, freq_mask=freq_mask,
+                     welch_fun=welch)
+                     for d in x_splits)
+
+    # Combining/reshaping to original data shape
+    psds = np.concatenate(f_psd, axis=0)
+    psds = psds.reshape(np.hstack([dshape, -1]))
+    return psds, freqs
+
+
+@verbose
+def psd_welch(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None, n_fft=256,
+              n_overlap=0, picks=None, proj=False, n_jobs=1, verbose=None):
+    """Compute the power spectral density (PSD) using Welch's method.
+
+    Calculates periodigrams for a sliding window over the
+    time dimension, then averages them together for each channel/epoch.
+
+    Parameters
+    ----------
+    inst : instance of Epochs or Raw or Evoked
+        The data for PSD calculation
+    fmin : float
+        Min frequency of interest
+    fmax : float
+        Max frequency of interest
+    tmin : float | None
+        Min time of interest
+    tmax : float | None
+        Max time of interest
+    n_fft : int
+        The length of the tapers ie. the windows. The smaller
+        it is the smoother are the PSDs. The default value is 256.
+        If ``n_fft > len(inst.times)``, it will be adjusted down to
+        ``len(inst.times)``.
+    n_overlap : int
+        The number of points of overlap between blocks. Will be adjusted
+        to be <= n_fft.
+        bandwidth : float
+        The bandwidth of the multi taper windowing function in Hz.
```

Comment:
Good catch, I missed it because of the indentation, will remove it
On Jan 7, 2016 8:58 AM, "Eric Larson" <notifications@github.com> wrote:

> In mne/time_frequency/psd.py
> <https://github.com/mne-tools/mne-python/pull/2710#discussion_r49080602>:
>
> > +    fmax : float
> > +        Max frequency of interest
> > +    tmin : float | None
> > +        Min time of interest
> > +    tmax : float | None
> > +        Max time of interest
> > +    n_fft : int
> > +        The length of the tapers ie. the windows. The smaller
> > +        it is the smoother are the PSDs. The default value is 256.
> > +        If ``n_fft > len(inst.times)``, it will be adjusted down to
> > +        ``len(inst.times)``.
> > +    n_overlap : int
> > +        The number of points of overlap between blocks. Will be adjusted
> > +        to be <= n_fft.
> > +        bandwidth : float
> > +        The bandwidth of the multi taper windowing function in Hz.
>
> ? this param isn't indented properly, I don't see it above, and I don't
> think it's actually used in the welch method anyway...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2710/files#r49080602>.
>